src/analysis/inner_product_space/reproducing_kernel.lean

```diff
@@ -0,0 +1,257 @@
+/-
+Copyright (c) 2022 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Shing Tak Lam
+-/
+
+import analysis.inner_product_space.projection
+import analysis.inner_product_space.positive
+
+/-!
+# Reproducing Kernel Hilbert Space
+
+A Hilbert space `H` is called a reproducing kernel Hilbert space (RKHS) if it is a vector subspace
+of `X → V` where `V` is a Hilbert space, and the evaluation map `H → V` at each `X` is continuous.
+
+## Main definitions
+
+* `rkhs 𝕜 X V H` - `H` is a RKHS, where `𝕜` is the scalar field, `X` is the domain, `V` is the
+  Hilbert space.
+* `rkhs.kernel` - the kernel of a RKHS, as a continuous linear map `V →L[𝕜] V`
+* `rkhs.eval'` - For `V = 𝕜`, by the Riesz representation theorem, the evaluation map can be
+  represented as the inner product with a point in `H`, which we call the `eval'`
+* `rkhs.scalar_kernel` - For `V = 𝕜`, we can represent the kernel of `H` as an inner product with a
+  point in `H`, which we call the `scalar_kernel`
+* `rkhs.eval'_span` - The span of the `eval'`s as a subspace of `H`.
+
+## Main statements
+
+* `rkhs.eval'_span_topological_closure_eq_top` - the topological closude of the span of the
+  `eval's` is the whole space `H`.
+* `rkhs.tendsto_nhds_apply` - Convergence in `H` implies pointwise convergence.
+
+## Notes
+
+Due to the statement of the Riesz representation theorem in mathlib, the order of `x` and `y` in
+the definition of the `scalar_kernel` and the representation of it as an inner product of
+`eval'`s is reversed. I opted to keep the statement simpler, but this leaves the order
+reversed.
+
+## Tags
+
+Reproducing kernel Hilbert space, RKHS
+-/
+
+noncomputable theory
+
+open_locale inner_product topological_space
+open inner_product_space filter
+
+variables (𝕜 : Type*) [is_R_or_C 𝕜]
+variables (X : Type*)
+variables (V : Type*) [inner_product_space 𝕜 V]
+
+/--
+A reproducing kernel Hilbert space (RKHS) is a vector subspace of `X → V` where evaluation at a
+point is continuous.
+-/
+class rkhs (H : Type*) [inner_product_space 𝕜 H] extends fun_like H X (λ _, V) :=
+(add_apply' : ∀ {f g : H} {x : X}, (f + g) x = f x + g x)
+(smul_apply' : ∀ {c : 𝕜} {f : H} {x : X}, (c • f) x = c • f x)
+(continuous_eval' : ∀ (x : X), continuous (λ (f : H), f x))
```

Comment:
This design seems dangerous: `fun_like` doesn't come with any addition, which means that the addition that appears here comes from the `[inner_product_space 𝕜 H]` instance, which might be completely unrelated to pointwise addition. Also, in many cases this means that you would wind up with two `has_add` instances on `H`. I'm not sure what the right solution to this is at this point, but it does mean that the design should be rethought.

Reply:
I see... I think there is a branch somewhere that people have added typeclasses for saying when a `fun_like` class has pointwise ops, with the same design as what I have here, but that hasn't been merged to master. 

I suppose an alternative would be to define it as a predicate on a submodule of `X -> k`. The issue with that though is that comes with a topology (the product topology) which might not be the one that I want for arbitrary inner products, so I'll need an alias that preserves the module and pointwise structure, but not the topological structure.

Reply:
https://github.com/leanprover-community/mathlib/blob/0a8652754854bb42b93e7c951cd748fb6a3a4330/src/algebra/group/fun_like.lean

Reply:
I agree that a type class like this would probably be the way to go, in fact something like this is probably going to be required at some point to have operator algebras.

Reply:
(On second thought, you're right that my concern about having multiple additions is probably unfounded.)

Reply:
(On second thought, you're right that my concern about multiple addition instances is probably unfounded.)